polaris-react/src/components/BulkActions/BulkActions.tsx

```diff
@@ -51,408 +60,314 @@ export interface BulkActionsProps {
-    }
-
-    this.setState(({popoverVisible}) => ({
-      popoverVisible: !popoverVisible,
-    }));
-  };
+      return memo.concat(action);
+    },
+    [] as (BulkAction | MenuGroupDescriptor)[],
+  );
 
-  private handleMeasurement = (width: number) => {
-    const {measuring} = this.state;
-    if (measuring) {
-      this.promotedActionsWidths.push(width);
-    }
+  const hiddenPromotedSection = {
+    items: mergedHiddenPromotedActions,
   };
-}
-
-function instanceOfBulkActionListSectionArray(
-  actions: (BulkAction | BulkActionListSection)[],
-): actions is BulkActionListSection[] {
-  const validList = actions.filter((action: any) => {
-    return action.items;
-  });
-
-  return actions.length === validList.length;
-}
 
-function instanceOfBulkActionArray(
-  actions: (BulkAction | BulkActionListSection)[],
-): actions is BulkAction[] {
-  const validList = actions.filter((action: any) => {
-    return !action.items;
-  });
+  const allHiddenActions = actions
+    ? actions
+        .filter((action) => action)
+        .map(
+          (
+            action: BulkAction | MenuGroupDescriptor | BulkActionListSection,
+          ) => {
+            if (instanceOfBulkActionListSection(action)) {
+              return {items: [...action.items]};
+            } else if (instanceOfMenuGroupDescriptor(action)) {
+              return {items: [...action.actions]};
+            }
+            return {items: [action]};
+          },
+        )
+    : [];
+
+  const activator = (
+    <BulkActionButton
+      disclosure
+      showContentInButton={!promotedActionsMarkup}
+      onAction={togglePopover}
+      content={activatorLabel}
+      disabled={disabled}
+      indicator={isNewBadgeInBadgeActions(actionSections)}
+      size={buttonSize}
+    />
+  );
 
-  return actions.length === validList.length;
-}
+  const actionsMarkup =
+    allHiddenActions.length > 0 ? (
+      <Popover
+        active={popoverActive}
+        activator={activator}
+        preferredAlignment="right"
+        onClose={togglePopover}
+      >
+        <ActionList
+          sections={[hiddenPromotedSection, ...allHiddenActions]}
+          onActionAnyItem={togglePopover}
+        />
+      </Popover>
+    ) : null;
+
+  const measurerMarkup = (
+    <BulkActionsMeasurer
+      promotedActions={promotedActions}
+      disabled={disabled}
+      buttonSize={buttonSize}
+      handleMeasurement={handleMeasurement}
+    />
+  );
 
-function instanceOfMenuGroupDescriptor(
-  action: MenuGroupDescriptor | BulkAction,
-): action is MenuGroupDescriptor {
-  return 'title' in action;
-}
+  const bulkActionsClassNames = classNames(styles.BulkActions);
 
-export const BulkActions = forwardRef(function BulkActions(
-  props: BulkActionsProps,
-  ref,
-) {
-  const i18n = useI18n();
-
-  return <BulkActionsInner {...props} i18n={i18n} innerRef={ref} />;
+  return (
+    <div className={bulkActionsClassNames} style={width ? {width} : undefined}>
```

Comment:
```suggestion
    <div className={styles.BulkActions} style={width ? {width} : undefined}>
```

Light suggestion to just use `styles.BulkActions` directly here and remove `bulkActionsClassNames` variable higher up since classnames aren't being combined.